### ISSUE TITLE
shared_examples inconsistencies

### ISSUE THREAD

**Issue by fabioxgn** [rubocop-hq/rubocop-rspec#1008]
I've update this gem to v1.43.1 and it started catching `RSpec/FilePath` in my `spec/support/shared_examples` folder, but I found the behavior confusing and inconsistent, some examples:

```ruby
# spec/support/shared_examples/my_class.rb
shared_examples_for MyClass do
   ...
end
```

The code above breaks with `RSpec/FilePath`, it expects `my_class_spec.rb` instead of just `my_class`, but the [rspec docs](https://relishapp.com/rspec/rspec-core/docs/example-groups/shared-examples) and [better specs](https://www.betterspecs.org/#shared) doesn't mention this convention, where does it come from? Looks like a bug to me.

But this is just one problem, if you have this instead:

```ruby
# spec/support/shared_examples/my_class.rb
shared_examples_for "my class" do
   # lots of ofenses
end
```

The code above doesn't break the `RSpec/FilePath` cop but also **doesn't break at cops at all** in the code, this file is just ignored completly by rubocop. If I change `"my class"` to `MyClass` then it catches all the ofenses and along with the `RSpec/FilePath`.

The third problem is when you have a module, like this:

```ruby
# spec/support/shared_examples/my_module_my_class.rb
shared_examples_for MyModule::MyClass do
   ...
end
```

It breaks with `RSpec/FilePath`, as it expects the file to be in `spec/support/shared_examples/my_module/my_class_spec.rb`, but if you move it there, RSpec fails to run:

```
NoMethodError:
  undefined method `shared_examples_for' for main:Object
```

Keeping it short:

- Where does the `_spec.rb` file name convention from `shared_examples` come from? Does it make sense to just break this cop when the shared example is declared using a class and not break when it is a string?
- Why the cops don't catch ofenses when you declared the shared example with a string?
- The module problem is some RSpec config I am missing?

**Comment by pirj** [comment]
> also doesn't break at cops at all in the code

This is not exactly true to my best knowledge. Can you provide some examples of what is being ignored by cops inside a shared example group with a literal docstring? It makes sense to fix that if it is the case.

**Comment by fabioxgn** [comment]
> > also doesn't break at cops at all in the code
> 
> This is not exactly true to my best knowledge. Can you provide some examples of what is being ignored by cops inside a shared example group with a literal docstring? It makes sense to fix that if it is the case.

Sorry, I edited the issue, the files where in `rubocop_todo.yml`, my bad 🤦‍♂️ .

**Comment by pirj** [comment]
> doesn't mention this convention, where does it come from? Looks like a bug to me

You're correct, it's a bug. There's no convention that RuboCop RSpec is trying to push through here.

**Comment by pirj** [comment]
> RSpec fails to run:

```
NoMethodError:
  undefined method `shared_examples_for' for main:Object
```

I guess `disable_monkey_patching!` is turned on in `spec_helper.rb`, isn't it?
If not, I'd love to know what is causing this. RSpec should handle shared specs disregarding the file name they are in.

**Comment by fabioxgn** [comment]
> > RSpec fails to run:
> 
> ```
> NoMethodError:
>   undefined method `shared_examples_for' for main:Object
> ```
> 
> I guess `disable_monkey_patching!` is turned on in `spec_helper.rb`, isn't it?
> If not, I'd love to know what is causing this. RSpec should handle shared specs disregarding the file name they are in.

Yes, it is! Changing to `RSpec.shared_examples_for` solved it, but why does it work when the file is in `spec/support/shared_examples` and not in `spec/supports/shared_examples/my_module`? 🤔

**Comment by pirj** [comment]
```ruby
RSpec.describe "my class" do
   ...
end
```
> The code above doesn't expect the file to end with _spec.rb, shouldn't it be consistent?

It's hard to infer the file name for an arbitrary docstring, so the cop ignores non-constants.

When a constant is quoted, unfortunately, the cops ignores it as well.
You can help RuboCop RSpec by improving the cop's detection just [like `RSpec/DescribeClass` does](https://github.com/rubocop-hq/rubocop-rspec/blob/master/lib/rubocop/cop/rspec/describe_class.rb#L56). It's an easy task, and I'll be happy to help along the way.

**Comment by fabioxgn** [comment]
> ```ruby
> RSpec.describe "my class" do
>    ...
> end
> ```
> 
> > The code above doesn't expect the file to end with _spec.rb, shouldn't it be consistent?
> 
> It's hard to infer the file name for an arbitrary docstring, so the cop ignores non-constants.
> 
> When a constant is quoted, unfortunately, the cops ignores it as well.
> You can help RuboCop RSpec by improving the cop's detection just [like `RSpec/DescribeClass` does](https://github.com/rubocop-hq/rubocop-rspec/blob/master/lib/rubocop/cop/rspec/describe_class.rb#L56). It's an easy task, and I'll be happy to help along the way.

You mean, improve the cop to match the file name when using a string? Like match `shared_examples "my clever name" to `shared_examples/my_clever_name.rb`?

**Comment by pirj** [comment]
I don't think insisting on `my_clever_name.rb` is a good idea, but to my point of view
```ruby
RSpec.describe 'MyNamespace::MyClass' do
```
should be treated the same way as
```ruby
RSpec.describe MyNamespace::MyClass do
```
